### ISSUE TITLE
Assignment Expressions

### ISSUE THREAD

**Issue by da-dada** [PyCQA/pylint#3865]
if (init := getattr(inst, 'init', None)) and callable(init): gives E0601 Using variable 'init' before assignment and E1102 'init' is not callable ('inst' is a given instance and 'init' a method of it)

**Comment by dbaty** [comment]
Here is a reformatted snippet:

```python

# pylint: disable=missing-module-docstring, missing-class-docstring, missing-function-docstring
# pylint: disable=too-few-public-methods
# pylint: disable=invalid-name

class C:
    def init(self):
        pass

c = C()
if init := getattr(c, 'init', None) and callable(init):
    pass
```

pylint says:

```
test.py:10:49: E0601: Using variable 'init' before assignment (used-before-assignment)
```

In fact, Python 3.8 fails with a NameError:

```
$ python ~/test.py
Traceback (most recent call last):
  File "test.py", line 10, in <module>
    if init := getattr(c, 'init', None) and init():
NameError: name 'init' is not defined 
```

So I don't think that pylint output is wrong. I am not yet too familiar with the walrus operator, so I may be missing something. Perhaps this syntax works as of Python 3.9?

As for the E1102 warning (`not-callable`) that you report, you'll need to provide further details. Your snippet does not call `init`, so I don't see how pylint could report `not-callable` (and, as show above, it clearly doesn't). Also, note that pylint sometimes cannot guess if attributes fetched through `getattr` are of a particular type. This may be your issue on your real code (not the snippet you provided).

**Comment by da-dada** [comment]
Python 3.8 works fine and I know that a static linter can’t check all dynamic features
the Assignment Expressions is described in PEP 572 and what you seem to miss is

‘The := operator groups more tightly than a comma in all syntactic positions where it is legal, but less tightly than all other operators’

I gave the example ‘if (init := getattr(inst, 'init', None)) and callable(init):’ and accordingly the brackets
are of substance!!!!!!!!!!! the E1102 is a sequel and you may safely forget about it

I’m very pleased with pylint and I just like to help improving it

**Comment by dbaty** [comment]
Indeed, I forgot the parentheses. A reproducible snippet is hence:

```
# pylint: disable=missing-module-docstring

if (a := lambda: 2) and a():
    print("ok")
```

pylint output (master branch, as of 2020-09-29):

```
pylint ~/test.py
************* Module test
test.py:3:24: E0601: Using variable 'a' before assignment (used-before-assignment)
```

pylint is indeed wrong, here.

**Comment by hippo91** [comment]
@dbaty thanks for the snippet. Thanks to it i can reproduce the bug.
